pkg/engine/engine_test.go

```diff
@@ -96,3 +110,156 @@ func TestUnmarshalTaskBuild(t *testing.T) {
+		// get the task back from the database and check final state
+		taskFromDB, err := engine.store.Get(testTask.ID)
+		require.NoError(t, err)
+		require.Equal(t, task.StateComplete, taskFromDB.State().State)
+	})
+
+	t.Run("Stall case", func(t *testing.T) {
+		testTask := createTestTask(t, Stall)
+		require.NoError(t, engine.queue.Push(testTask))
+		time.Sleep(70 * time.Second) // need to give it a time to finish because it is asynchronous function.
+		// get the task back from the database and check final state
+		taskFromDB, err := engine.store.Get(testTask.ID)
+		require.NoError(t, err)
+		require.Equal(t, task.StateCanceled, taskFromDB.State().State)
+	})
+}
+
+// testcase could be one of these: 'ok' and 'stall'
+// OK will execute right away and stall will cause deadline exceed.
+func createTestTask(t *testing.T, testcase Testcase) *task.Task {
+	cmp := api.Composition{
+		Global: api.Global{
+			Builder:        "exec:go",
+			Plan:           "placebo",
+			Case:           string(testcase),
+			TotalInstances: 1,
+			BuildConfig: map[string]interface{}{
+				"go_proxy_mode": "direct",
+			},
+			Runner: "local:exec",
+			RunConfig: map[string]interface{}{
+				"enabled": true,
+			},
+		},
+		Groups: []*api.Group{
+			{
+				ID:        "test",
+				Instances: api.Instances{Count: 1},
+			},
+		},
+	}
+
+	manifest := api.TestPlanManifest{
+		Name: "placebo",
+		Builders: map[string]config.ConfigMap{
+			"exec:go": {},
+		},
+		Runners: map[string]config.ConfigMap{
+			"local:exec": {},
+		},
+		TestCases: []*api.TestCase{
+			{
+				Name:      string(testcase),
+				Instances: api.InstanceConstraints{Minimum: 1, Maximum: 100},
+				Parameters: map[string]api.Parameter{
+					"param4": {
+						Type:    "string",
+						Default: "value4:default:manifest",
+					},
+				},
+			},
+		},
+	}
+	basedir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return nil
+	}
+
+	plandir := filepath.Join(basedir, "plan")
+	err = copy.Copy("../../plans/placebo", plandir)
+	if err != nil {
+		return nil
+	}
+	t.Cleanup(func() {
+		os.RemoveAll(basedir)
+	})
+
+	id := xid.New().String()
+	taskData := &task.Task{
+		Type:        task.TypeRun,
+		Version:     0,
+		Priority:    0,
+		Plan:        "placebo",
+		Case:        string(testcase),
+		ID:          id,
+		Runner:      "local:exec",
+		Composition: cmp,
+		Input: RunInput{
+			RunRequest: &api.RunRequest{
+				CreatedBy:   api.CreatedBy{User: "test-user"},
+				BuildGroups: []int{0},
+				Composition: cmp,
+				Manifest:    manifest,
+			},
+
+			Sources: &api.UnpackedSources{BaseDir: basedir, PlanDir: plandir},
+		},
+		States: []task.DatedState{
+			{
+				State:   task.StateScheduled,
+				Created: time.Now().UTC(),
+			},
+		},
+		CreatedBy: task.CreatedBy(api.CreatedBy{User: "test-user"}),
+	}
+
+	jsonData, err := json.Marshal(taskData)
+
+	if err != nil {
+		return nil
+	}
+
+	newTask, err := UnmarshalTask(jsonData)
+
+	if err != nil {
+		return nil
+	}
+
+	return newTask
+}
```

Comment:
Testground's integration testing is quite messy at the moment, see https://github.com/testground/testground/issues/1368 for references.

While we wait to focus on the integration testing story,
Shell scripts have the benefit of introducing less legacy code (this is ~100+ lines of hardcoded configuration that already exists in the `plan` folder). Could you move this test to integration tests?

In the end, it looks like we are "just" testing the outcome:

```
testground run placebo:ok test
assert task status === "canceled"

testground run placebo:ok test
assert task status === "canceled"
```